<a href="https://colab.research.google.com/github/muinul-official/Data-Mining-G13/blob/hannah-feature-engineering/EDA_%26_Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# WIE3007 Data Mining Group Project
##EDA & Feature Engineering

* Name: Hannah bt Arham Rahimy
* Matrics Number : 22002434
* Role: Data Cleaning, Encoding, Feature Engineering, AI-based Feature Extraction



###1) Load Dataset

In [1]:
import pandas as pd
churn = pd.read_csv("ecommerce_churn_dataset.csv")
transactions = pd.read_csv("ecommerce_transactions.csv")
chats = pd.read_csv("ecommerce_chat_logs.csv")
reviews = pd.read_csv("ecommerce_reviews.csv")

print("churn:", churn.shape)
print("transactions:", transactions.shape)
print("chats:", chats.shape)
print("reviews:", reviews.shape)

churn.head()

churn: (1200, 30)
transactions: (6396, 8)
chats: (1200, 5)
reviews: (1200, 5)


customer_id  age  gender  location membership_tier  account_age_days  \
0  CUST000001   58    Male     Urban            Gold               480   
1  CUST000002   18    Male     Rural        Platinum               382   
2  CUST000003   73  Female     Rural          Silver                 9   
3  CUST000004   61   Other     Urban          Silver               184   
4  CUST000005   28   Other  Suburban          Bronze               336   

   recency  frequency  monetary_value  avg_order_value  ...  \
0        0          2           88.15            44.08  ...   
1        0          2          245.31           122.66  ...   
2        9          1           18.85            18.85  ...   
3        0          6          126.75            21.12  ...   
4        0          3           67.16            22.39  ...   

                                         review_text  review_rating  \
0                 Average purchase. No major issues.              3   
1                    It's fine. Nothing exceptional.              3   
2  The worst purchase I've made. Don't buy this p...              2   
3  Product doesn't work as advertised. Very frust...              2   
4  Wonderful product. Fast delivery. Highly recom...              4   

   review_sentiment_score  review_length  sentiment_score_avg  \
0                   0.520             34                0.533   
1                   0.594             31                0.499   
2                   0.055             53                0.055   
3                   0.110             53                0.099   
4                   0.714             51                0.796   

   text_engagement_score total_spent_per_day  order_frequency_per_month  \
0                  0.445            0.183264                   0.117647   
1                  0.440            0.640496                   0.145631   
2                  0.610            1.885000                   0.769231   
3                  0.635            0.685135                   0.841121   
4                  0.585            0.199288                   0.245902   

   churn  churn_probability  
0      0              0.330  
1      0              0.436  
2      0              0.489  
3      1              0.577  
4      0              0.158  

[5 rows x 30 columns]

###2) EDA

In [4]:
churn.columns

Index(['customer_id', 'age', 'gender', 'location', 'membership_tier',
       'account_age_days', 'recency', 'frequency', 'monetary_value',
       'avg_order_value', 'total_orders', 'days_since_last_login',
       'cart_abandonment_rate', 'email_open_rate', 'support_tickets',
       'frustration_index', 'chat_log', 'chat_sentiment_score', 'chat_length',
       'has_negative_keywords', 'review_text', 'review_rating',
       'review_sentiment_score', 'review_length', 'sentiment_score_avg',
       'text_engagement_score', 'total_spent_per_day',
       'order_frequency_per_month', 'churn', 'churn_probability'],
      dtype='object')

In [5]:
churn.head()

customer_id  age  gender  location membership_tier  account_age_days  \
0  CUST000001   58    Male     Urban            Gold               480   
1  CUST000002   18    Male     Rural        Platinum               382   
2  CUST000003   73  Female     Rural          Silver                 9   
3  CUST000004   61   Other     Urban          Silver               184   
4  CUST000005   28   Other  Suburban          Bronze               336   

   recency  frequency  monetary_value  avg_order_value  ...  \
0        0          2           88.15            44.08  ...   
1        0          2          245.31           122.66  ...   
2        9          1           18.85            18.85  ...   
3        0          6          126.75            21.12  ...   
4        0          3           67.16            22.39  ...   

                                         review_text  review_rating  \
0                 Average purchase. No major issues.              3   
1                    It's fine. Nothing exceptional.              3   
2  The worst purchase I've made. Don't buy this p...              2   
3  Product doesn't work as advertised. Very frust...              2   
4  Wonderful product. Fast delivery. Highly recom...              4   

   review_sentiment_score  review_length  sentiment_score_avg  \
0                   0.520             34                0.533   
1                   0.594             31                0.499   
2                   0.055             53                0.055   
3                   0.110             53                0.099   
4                   0.714             51                0.796   

   text_engagement_score total_spent_per_day  order_frequency_per_month  \
0                  0.445            0.183264                   0.117647   
1                  0.440            0.640496                   0.145631   
2                  0.610            1.885000                   0.769231   
3                  0.635            0.685135                   0.841121   
4                  0.585            0.199288                   0.245902   

   churn  churn_probability  
0      0              0.330  
1      0              0.436  
2      0              0.489  
3      1              0.577  
4      0              0.158  

[5 rows x 30 columns]

In [16]:
#target variable understanding
churn['churn'].value_counts()

churn
0    925
1    275
Name: count, dtype: int64

In [10]:
#check if customer_id exists in all datasets
transactions.columns

Index(['customer_id', 'transaction_date', 'amount', 'product_category',
       'quantity', 'discount_applied', 'payment_method', 'shipping_method'],
      dtype='object')

In [11]:
chats.columns

Index(['customer_id', 'chat_log', 'chat_sentiment_score', 'chat_length',
       'has_negative_keywords'],
      dtype='object')

In [12]:
reviews.columns

Index(['customer_id', 'review_text', 'review_rating', 'review_sentiment_score',
       'review_length'],
      dtype='object')

In [25]:
#check dataset shape & missing values
transactions.shape, transactions.isnull().sum()

((6396, 8),
 customer_id         0
 transaction_date    0
 amount              0
 product_category    0
 quantity            0
 dtype: int64)

In [26]:
chats.shape, chats.isnull().sum()

((1200, 5),
 customer_id              0
 chat_log                 0
 chat_sentiment_score     0
 chat_length              0
 has_negative_keywords    0
 dtype: int64)

In [27]:
reviews.shape, reviews.isnull().sum()

((1200, 5),
 customer_id               0
 review_text               0
 review_rating             0
 review_sentiment_score    0
 review_length             0
 dtype: int64)

#### EDA Summary

The main churn dataset contains 1,200 customer-level records with a binary churn target, where 925 customers are non-churned and 275 are churned, indicating moderate class imbalance.

Supporting datasets include transaction history (6396 records), customer support chat logs (1200), and product reviews (1200). Transaction data is recorded at transaction level, while chat and review datasets are already aggregated at customer level.

All datasets share a common `customer_id`, enabling integration across data sources. Initial inspection shows no missing values in key variables, indicating good data quality for feature engineering and modeling.


###3) Data Cleaning & Preprocessing

In [18]:
# correcting data types
transactions['amount'] = pd.to_numeric(transactions['amount'], errors='coerce')

In [20]:
# check duplicate
transactions.duplicated().sum()

np.int64(0)

In [22]:
chats.duplicated().sum()

np.int64(0)

In [23]:
reviews.duplicated().sum()

np.int64(0)

In [21]:
transactions.describe()

amount     quantity
count  6396.000000  6396.000000
mean     69.715044     3.020951
std     131.189101     1.417924
min       0.400000     1.000000
25%      14.812500     2.000000
50%      33.320000     3.000000
75%      74.935000     4.000000
max    3683.010000     5.000000

####Data Cleaning & Preprocessing Summary :

Basic preprocessing was applied, including data type correction, duplicate checks,
and sanity inspection of numerical values. No duplicate records were identified, and transaction
amounts were confirmed to be within reasonable ranges. The datasets were found to be clean and
suitable for feature engineering.
